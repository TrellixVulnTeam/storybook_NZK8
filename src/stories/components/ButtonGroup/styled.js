// @flow

import styled from 'styled-components';
import { colors } from 'utils/StyleGuide';
import scrumIcon from 'assets/commonIcons/scrumIcon.svg';
import calendarIcon from './img/calendarIcon.svg';
import listIcon from './img/listIcon.svg';

/**
 * Determine the button group's size based on the "size" prop
 * @param {string} size
 * @return {string}
 */
function getSize(size) {
  switch (size) {
    case 'xs':
      return { width: 36, height: 30, font: 10, vpad: 0, hpad: 4 };
    case 'sm':
      return { width: 36, height: 36, font: 11, vpad: 0, hpad: 8 };
    case 'lg':
      return { width: 60, height: 48, font: 20, vpad: 0, hpad: 16 };
    default:
      return { width: 80, height: 36, font: 12, vpad: 0, hpad: 12 };
  }
}

export const Wrapper = styled.div`
  margin: 0;
  border-radius: 6px;
  display: ${p => (p.flex ? 'flex' : 'inline-flex')};
  height: ${p => getSize(p.size).height}px;

  [type='radio'] {
    display: none;
  }

  .radio-option {
    min-width: ${p => getSize(p.size).width}px;
    padding: ${p => getSize(p.size).vpad}px ${p => getSize(p.size).hpad}px;
    text-align: center;
    cursor: pointer;
    border: 1px solid #ccc;
    transition: 100ms;
    flex-grow: ${p => p.flex && 1};

    span {
      line-height: ${p => getSize(p.size).height}px;
      font-size: ${p => getSize(p.size).font}px;
    }

    &.disabled {
      background: #ddd;

      span {
        color: #aaa;
      }
    }
    svg {
      fill: ${colors.silver};
    }
  }

  .radio-option:hover:not(.disabled):not(.selected) {
    background: ${colors.white};
    transition: 100ms;
    .option-label {
      color: ${colors.brightBlue};
    }

    svg {
      fill: ${colors.brightBlue};
    }
  }

  .radio-option.selected {
    transition: 100ms;
    background: ${colors.iceBlue};

    &.disabled {
      background-image: linear-gradient(251deg, #888, #777);
    }

    span {
      color: ${colors.brightBlue};
    }

    svg {
      fill: ${colors.brightBlue};
    }
  }

  .tooltip-option {
    flex-grow: ${p => p.flex && 1};

    &:first-child {
      .radio-option {
        border-radius: 6px 0 0 6px;
      }
    }

    &:not(:first-child):not(:last-child) {
      .radio-option {
        border-left: none;
      }
    }

    &:last-child {
      .radio-option {
        border-radius: 0 6px 6px 0;
        border-left: none;
      }
    }
  }

  .option-label {
    color: ${colors.silver};
    font-size: 12px;
  }
`;

export const CalendarIcon = styled(calendarIcon)`
  width: 16px;
  height: 16px;
  margin-top: 10px;
`;

export const ListIcon = styled(listIcon)`
  width: 16px;
  height: 16px;
  margin-top: 10px;
`;

export const ScrumIcon = styled(scrumIcon)`
  width: 16px;
  height: 16px;
  margin-top: 10px;
`;
