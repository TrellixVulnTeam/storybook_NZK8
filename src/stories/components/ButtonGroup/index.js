// @flow

import React from 'react';
import Tooltip from 'components/Tooltip';
import { Wrapper, CalendarIcon, ListIcon, ScrumIcon } from './styled';

type Props = {
  options: Array<Object>,
  tooltip?: string,
  value: string | Array<string>,
  onChange?: Function,
  theme?: 'blue-gradient' | 'red-gradient',
  size?: 'xs' | 'sm' | 'md' | 'lg' | 'xl',
  multiSelect?: boolean,
  disabled?: boolean,
  flex?: boolean,
};

/**
 * Render a radio button group
 * @param {Object} props
 * @return {React$Node}
 */
export default (props: Props) => (
  <Wrapper clearable={false} {...props}>
    {(() => {
      const value = Array.isArray(props.value) ? props.value : [props.value];
      return props.options.map(function(option) {
        return (
          <Tooltip
            className="tooltip-option"
            id={option.tooltip}
            value={option.tooltip}
            effect="solid"
            key={option.value}
          >
            {/* eslint-disable-next-line jsx-a11y/label-has-associated-control */}
            <label
              className={`radio-option ${option.disabled || props.disabled ? 'disabled ' : ''}${
                value.includes(option.value) ? 'selected' : ''
              }`}
            >
              <input
                type="radio"
                className="radio-option-value"
                name="option"
                value={option.value}
                onChange={ev => {
                  if (props.disabled) {
                    return;
                  }

                  // eslint-disable-next-line no-shadow
                  const value = ev ? ev.target.value : '';
                  if (typeof props.onChange === 'function') {
                    let result;
                    if (props.multiSelect) {
                      // eslint-disable-next-line no-shadow
                      const availableOptions = props.options.map(option => {
                        return option.value;
                      });
                      const selectedOptions = Array.isArray(props.value)
                        ? props.value.slice()
                        : [props.value];
                      const optionsObj = {};

                      // eslint-disable-next-line no-shadow
                      availableOptions.forEach(option => {
                        optionsObj[option] = selectedOptions.includes(option);
                      });

                      if (optionsObj.hasOwnProperty(value)) {
                        optionsObj[value] = !optionsObj[value];
                      } else {
                        optionsObj[value] = true;
                      }

                      // eslint-disable-next-line no-shadow
                      result = Object.keys(optionsObj).filter(value => {
                        return optionsObj[value];
                      });
                    } else {
                      result = value;
                    }

                    props.onChange(result);
                  }
                }}
              />
              {(() => {
                if (option.label) {
                  return <span className="option-label">{option.label}</span>;
                }
                if (option.image === 'calendar') {
                  return <CalendarIcon />;
                }
                if (option.image === 'list') {
                  return <ListIcon />;
                }
                if (option.image === 'scrum') {
                  return <ScrumIcon />;
                }
              })()}
            </label>
          </Tooltip>
        );
      });
    })()}
  </Wrapper>
);
