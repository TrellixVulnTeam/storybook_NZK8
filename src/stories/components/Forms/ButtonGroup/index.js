// @flow

import React from 'react';
import { Field } from 'formik';
import ButtonGroupRaw from 'components/ButtonGroup';
import { Wrapper } from './styled';

/**
 * Render a react-select input, for use with
 * Formik's <Field> component or as a standalone component.
 * @param {Object} props
 * @return {React$Node}
 */
const ButtonGroup = (props: *) => {
  return (
    <ButtonGroupRaw
      value={props.field.value}
      onChange={value => {
        if (!Array.isArray(value) && typeof value !== 'string') {
          return;
        }

        props.form.setFieldValue(props.field.name, value);

        if (typeof props.onChange === 'function') {
          props.onChange(value);
        }
      }}
      {...props}
    />
  );
};

// NOTE: We intentionally extract className so we don't accidentally pass it into <Field />
export default ({ className, error, ...props }: *) => (
  <Wrapper error={error} {...props}>
    <Field {...props} component={ButtonGroup} />
  </Wrapper>
);
