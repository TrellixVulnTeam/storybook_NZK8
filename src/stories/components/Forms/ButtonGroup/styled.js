// @flow
/* eslint-disable import/prefer-default-export */

import styled from 'styled-components';

export const Wrapper = styled.div`
  display: ${p => (p.flex ? 'block' : 'inline-flex')};
`;
