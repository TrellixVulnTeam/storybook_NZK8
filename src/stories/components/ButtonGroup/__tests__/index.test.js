import React from 'react';
import { mount } from 'enzyme';
import ButtonGroup from 'components/ButtonGroup/index';

describe('<ButtonGroup />', () => {
  const mandatoryProps = {
    options: [
      {
        tooltip: 'tooltip',
        value: '0',
      },
    ],
    tooltip: 'tooltip',
    value: ['0', '1', '2'],
    onChange: jest.fn(),
  };

  it('should render correctly with mandatory props', () => {
    const sut = mount(<ButtonGroup {...mandatoryProps} />);

    expect(sut).toMatchSnapshot();
  });
});
