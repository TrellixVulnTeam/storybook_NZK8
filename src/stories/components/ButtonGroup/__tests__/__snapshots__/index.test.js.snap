// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonGroup /> should render correctly with mandatory props 1`] = `
<_default
  onChange={[MockFunction]}
  options={
    Array [
      Object {
        "tooltip": "tooltip",
        "value": "0",
      },
    ]
  }
  tooltip="tooltip"
  value={
    Array [
      "0",
      "1",
      "2",
    ]
  }
>
  <styled__Wrapper
    clearable={false}
    onChange={[MockFunction]}
    options={
      Array [
        Object {
          "tooltip": "tooltip",
          "value": "0",
        },
      ]
    }
    tooltip="tooltip"
    value={
      Array [
        "0",
        "1",
        "2",
      ]
    }
  >
    <div
      className="styled__Wrapper-sc-6sc1ip-0 fPqTDi"
      onChange={[MockFunction]}
      value={
        Array [
          "0",
          "1",
          "2",
        ]
      }
    >
      <_default
        className="tooltip-option"
        effect="solid"
        id="tooltip"
        key="0"
        value="tooltip"
      >
        <styled__Wrapper
          className="tooltip-option"
          data-for="tooltip"
          data-tip="tooltip"
          effect="solid"
          id="tooltip"
          value="tooltip"
        >
          <div
            className="tooltip-option styled__Wrapper-funpqb-0 cznzQQ"
            data-for="tooltip"
            data-tip="tooltip"
            id="tooltip"
            value="tooltip"
          >
            <label
              className="radio-option selected"
            >
              <input
                className="radio-option-value"
                name="option"
                onChange={[Function]}
                type="radio"
                value="0"
              />
            </label>
            <styled__TooltipWrapper
              display="true"
            >
              <div
                className="styled__TooltipWrapper-funpqb-1 gQtIVd"
                display="true"
              >
                <ReactTooltip
                  clickable={false}
                  delayShow={300}
                  effect="solid"
                  getContent={[Function]}
                  html={true}
                  id="tooltip"
                  insecure={true}
                  multiline={true}
                  place="top"
                  resizeHide={true}
                  type="dark"
                  wrapper="div"
                >
                  <div
                    className="__react_component_tooltip place-top type-dark"
                    data-id="tooltip"
                    id="tooltip"
                  >
                    null
                  </div>
                </ReactTooltip>
              </div>
            </styled__TooltipWrapper>
          </div>
        </styled__Wrapper>
      </_default>
    </div>
  </styled__Wrapper>
</_default>
`;
